### ISSUE TITLE
Various problems with push

### ISSUE THREAD

**Issue by jrosdahl** [ingydotnet/git-subrepo#69]
From documentation I get the impression that it should work to run `git subrepo push <subdir>` without specifying a subrepo branch, but I can't get it to work.

To reproduce what I'm trying to do, change

    git subrepo push bar subrepo/bar

to

    git subrepo push bar

in `test/subrepo-push.t`. The test case then fails like this:

```
test/subrepo-push.t .. 1/? #   Failed test 'push message is correct'
#   at test/subrepo-push.t line 45.
#     got: 'Subrepo 'bar' has no new commits to push.'
#   expected: 'Subrepo 'bar' pushed to '../../../tmp/upstream/bar' (master).'
test/subrepo-push.t .. 2/? #   Failed test 'File 'tmp/owner/bar/FooBar' exists'
#   at test/setup line 81.
test/subrepo-push.t .. Failed 2/6 subtests 
```

I guess it's supposed to work? If not, perhaps documentation could be improved?

Don't know if it's related, but another bug, I think, is that it doesn't work to push immediately after clone. This can be reproduced by removing the

```
(
  cd $OWNER/bar
  add-new-files bargy
  git push
)
```

part of `test/subrepo-push.t`. The test case then fails with

```
test/subrepo-push.t .. git-subrepo: 'subrepo/bar' is not a valid commit.
test/subrepo-push.t .. No subtests run 
```

instead of the expected

```
test/subrepo-push.t .. 1/? #   Failed test 'File 'tmp/owner/bar/bargy' exists'
```

I tried to make my own little standalone testcase, but it unfortunately only made me more confused. Here it is:

```
git init subproject
(
    cd subproject

    echo foo >foo
    git add foo
    git commit -m foo
)

git init superproject
(
    cd superproject

    # echo bar >bar
    # git add bar
    # git commit -m bar

    git subrepo clone ../subproject
    echo foo >>subproject/foo
    git commit -am foo

    git subrepo push subproject
)
```

The last command fails like this:

```
fatal: ambiguous argument 'none..HEAD': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
Subrepo 'subproject' has no new commits to push.
```

When uncommenting the three commented lines, the failure is

```
The 'git rebase' command failed during your push.
You will need to finish the push by hand. Follow these steps:

  # Resolve the rebase conflicts
  git checkout master
  git subrepo push subproject subrepo-push/subproject

To abort the push and reset back to where you started:

  git rebase --abort
  git checkout ORIG_HEAD
  git subrepo clean subproject

See 'git help subrepo' for more help.
```

but unless I'm missing something, there should be no conflicts.

I'm using Git version 2.1.0 and git-subrepo 5c38bbc8b48ef73d5a56f603b77436be79a5ed12.

**Comment by AndrewSmart** [comment]
>I guess it's supposed to work? If not, perhaps documentation could be improved?

You're quite right the tests need some additional work.  With your first question yes I think it's supposed to work; I'm not on my development machine so I can't dig into things at this time.

> I think, is that it doesn't work to push immediately after clone. 

At a glance I don't think your hypothesis is correct. By removing that push of bargy to $UPSTREAM/bar, I believe the subsequent `git subrepo pull --quiet bar` doesn't create the subrepo/bar branch as there was nothing new to pull from $UPSTREAM/bar, leading to the subsequent `git subrepo push bar subrepo/bar` failure of `git-subrepo: 'subrepo/bar' is not a valid commit.` Perhaps a helpful change would be to reword that failure message to indicate that 'branch' doesn't exist.

What could be done is `git subrepo branch <subdir>`, which will create subrepo/bar with the commits meant to be pushed, prior to the `git subrepo push bar subrepo/bar`.  In full, I think this should work, with your aforementioned removal of the bargy push:
```<bash>
message="$(
    cd $OWNER/foo
    git subrepo pull --quiet bar #Nothing will be pulled since you'd removed bargy
    git subrepo branch bar --quiet #Creates subrepo/bar with the commits meant to be pushed
    git subrepo push bar subrepo/bar
  )"
```
I hope that helps as a guide.  I think it helps to know in the current design, subrepo push is more simple than it could be (it expects subrepo/bar to be ready to push). I think Ingy wrote about that somewhere, but I don't see it in the current documentation.

**Comment by AndrewSmart** [comment]
> I tried to make my own little standalone testcase, but it unfortunately only made me more confused.

To make your own test, it's easiest to just copy an existing one, e.g. subrepo-push.t -> mytest.t, edit it (keep the header code, though you don't need to call clone-foo-and-bar since you're not using those), then run your standalone test case via `bash test/mytest.t` (you must run it from that parent directory due to the include in the header code, but I guess you can change that).

Running any subrepo command that fails with the verbosity flags (e.g. -v and -d) is quite helpful for knowing what goes wrong and how to fix it.

If you want Ingy's attention quick jump on IRC

**Comment by jrosdahl** [comment]
>> I think, is that it doesn't work to push immediately after clone.

> At a glance I don't think that hypothesis is correct.

I was probably a bit unclear, but I didn't have any hypothesis, I just (while fiddling around) made an observation that it doesn't work to push immediately after clone. :) From the point of view of an end user who hasn't learned the underlying model of git-subrepo yet, I would expect it to work (or result in a better error message like you suggested) to run

```
git-subrepo clone ...
# Make changes to subrepo directory
git commit
git-subrepo push ...
```

because of similarity with the ordinary Git commands and workflow.

>> I tried to make my own little standalone testcase, but it unfortunately only made me more confused.

> To make your own test, it's easiest to just copy an existing one,

Once again I was probably unclear, sorry about that. I wasn't trying to write a new git-subrepo test case; the "little standalone test case" (with the commented lines uncommented) is essentially what I ran when testing out git-subrepo the first time, trying to get "git-subrepo push" to work. I still haven't understood what the problem with my use case is. I'll dig deeper when I get time.

**Comment by AndrewSmart** [comment]
>> At a glance I don't think that hypothesis is correct.
> I was probably a bit unclear, but I didn't have any hypothesis,

You're right, conjecture/guess would have been a better word choice on my part.

I'll try to address everything else later today.

**Comment by AndrewSmart** [comment]
Ok, my earlier guess was wrong:
> ```<bash>
message="$(
    cd $OWNER/foo
    git subrepo pull --quiet bar #Nothing will be pulled since you'd removed bargy
    git subrepo branch bar --quiet #Creates subrepo/bar with the commits meant to be pushed
    git subrepo push bar subrepo/bar
  )"```

I need to look into this a bit more.

> I think it helps to know in the current design, subrepo push is more simple than it could be (it expects the subrepo to be ready to pushed). I think Ingy wrote about that somewhere, but I don't see it in the current documentation.

Ok, I was wrong here too. There are two meanings to the `git subrepo push <subdir> [<branch>]` command, depending on the presence of that optional branch argument.  This distinction is in the documentation.

A side effect of the `git subrepo pull --quiet bar` command is that the `subrepo/bar` branch is left when there was something to pull; but when there was nothing to pull the branch isn't created.  So dropping the branch name in `git subrepo push bar subrepo/bar`, will result in it creating a temporary branch during it's work, instead of trying to use what you specify. Sorry I was misled here.

I suppose things could be improved here, so that the commands don't have to be changed contingent on if or if not the pull was able to pull anything new.

**Comment by AndrewSmart** [comment]
> The last command fails like this:
```<bash>
fatal: ambiguous argument 'none..HEAD': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
Subrepo 'subproject' has no new commits to push.```


Ok I believe fixed this 'none' issue in this AndrewSmart/git-subrepo@1da1289714d6bc9763d953145c706dc1bc66eb8f branch. All tests work, and your test case is solved with this (though it still encounters the later rebase conflict).

**Comment by AndrewSmart** [comment]
> I still haven't understood what the problem with my use case is. I'll dig deeper when I get time.

I agree there should be no conflicts in that use case. It looks like it's because the `subrepo-push/subproject` branch created by `git subrepo push subproject` is incorrectly formed (it consists of just one commit, instead of one commit for each `echo foo >> foo` (this is a bug in the subrepo:branch() code).  This incorrectly formed branch causes the `git rebase refs/subrepo/subproject/fetch subrepo-push/subproject` to fail. I have no more time tonight, I hope this helps.

**Comment by grimmySwe** [comment]
I am currently working on introducing the git-subrepo into our new git repos. During this I have found the same "issue" as jrosdahl with the discrepancy of the subrepo pull/push and the regular pull/push.

Currently its much better to directly run ```git subrepo push bar``` as it seems to perform the pull automatically. If I run a ```git subrepo pull bar``` first I will need to specify a branch parameter to the push ```git subrepo push subrepo/bar``` and that is easy to forget.

As it seems that there are references created during the pull, wouldn't it be a good idea to default to these if they exist and you don't supply any other argument?

One side note is also that if I perform changes in my main repo and then pull in changes from a subrepo twice without pushing my own changes they will vanish without a trace...

**Comment by grimmySwe** [comment]
Added a separate issue #98 for:
"One side note is also that if I perform changes in my main repo and then pull in changes from a subrepo twice without pushing my own changes they will vanish without a trace..."

**Comment by grimmySwe** [comment]
With the new solution proposed in branch issue/98 from #98 I think the push issue is solved. Branch parameter for the push is now only there if you want to perform special operations on a branch before pushing it. Using push without branch should suffice in most regular cases.

@AndrewSmart Your fix https://github.com/AndrewSmart/git-subrepo/commit/1da1289714d6bc9763d953145c706dc1bc66eb8f is solved as well?

**Comment by jrosdahl** [comment]
Solved by fix of #98.
